### PR TITLE
Vector plots 2

### PR DESCRIPTION
**Pull request by pp-mo**
**WIP : not yet ready**

Followon work from #3096 
  * more review issues : see #3114
  * PLUS : tidy up + add some proper tests

~**PLEASE NOTE**~
~contents of `...tests/unit/analysis/cartography/test_gridcell_angles.py` are still mostly junk~
~Please await more commits before commenting !~
(now more sensible, though more tests still wanted)

### CODE REVIEW

**Comment by lbdreyer** [comment]
@pp-mo I think you also pushed this branch directly to upstream as we have a upstream/vector_plots_2d branch. I assume this was a mistake?

**Comment by pp-mo** [comment]
> I think you also pushed this branch directly to upstream as we have a upstream/vector_plots_2d branch. I assume this was a mistake?

I only did that when testing the auto-build of docs.
It actually wasn't this branch but something different with the same name !
(now removed).

So, not relevant to this.

**Comment by pp-mo** [comment]
Note: now rebased onto 'pin_dask', so we can hopefully have working tests (for now!).

This should look a lot more sensible.

However, it is still lacking some tests of the `grid_angles` method, like failure methods and some features.
In principle, it can use a grid with a rotated or even non-latlon coord system, but that quite possibly does not even work yet, and could even instead be removed ? Undecided.

Also, no tests yet for `rotate_grid_vectors` function.

**Comment by pp-mo** [comment]
hi @lbdreyer tests now passing : ready for a look-over ?

File: lib/iris/analysis/_grid_angles.py [hunk 1]
```diff
@@ -24,6 +24,7 @@
 
 import numpy as np
 
+import cartopy.crs as ccrs
 import iris
```
Comment (lbdreyer) [inline lib/iris/analysis/_grid_angles.py]: Potential changes to the `_3d_xyz_from_latlon` function
```
    Returns:
        xyz : (array, dtype=float64)
            cartesian coordinates on a unit sphere.  Dimension 0 maps x,y,z.
```
Clarify the last sentence? `Dimension 0 maps x,y,z`

`lon1` and `lat1` -> `lon_rad` and `lat_rad`? seems a little more informative

File: lib/iris/analysis/_grid_angles.py [hunk 2]
```diff
@@ -65,13 +66,13 @@ def _latlon_from_xyz(xyz):
     Returns:
 
         lonlat : (array)
-            spherical angles, of dims (2, <others>), in radians.
+            spherical angles, of dims (2, <others>), in degrees.
             Dim 0 maps longitude, latitude.
```
Comment (lbdreyer) [inline lib/iris/analysis/_grid_angles.py]: Here's this `Dim - maps longitude, latitde` again. I don't know what that means?

Comment (lbdreyer) [reply lib/iris/analysis/_grid_angles.py]: (I do prefer that this is working in degrees)

File: lib/iris/analysis/_grid_angles.py [hunk 3]
```diff
@@ -65,13 +66,13 @@ def _latlon_from_xyz(xyz):
     Returns:
 
         lonlat : (array)
-            spherical angles, of dims (2, <others>), in radians.
+            spherical angles, of dims (2, <others>), in degrees.
```
Comment (lbdreyer) [inline lib/iris/analysis/_grid_angles.py]: I do prefer that this is working in degrees to be consistent with the `_3d_xyz_from_latlon` function

Comment (pp-mo) [reply lib/iris/analysis/_grid_angles.py]: hi @lbdreyer 
I'll look into these now if you like, 
but in principle I had already deferred these "comments on the comments" to a follow-on : #3109, also linked from #3114.

I have been working on the follow-up, to address those remaining issues + add missing tests ...

**Comment by pp-mo** [comment]
Latest tries to improve docstrings, 
Is probably overkill (!)

**Comment by pp-mo** [comment]
Hi @lbdreyer :bouquet: :champagne: for checking the trigonometry code !
I'm hoping my latest addresses all your outstanding points.

**Comment by lbdreyer** [comment]
@pp-mo Thanks for this! The changes you've added all look fine so I'll squash and merge this in

**Pull request merged**